apisix/plugins/azure-functions.lua

```diff
@@ -0,0 +1,104 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+local core = require("apisix.core")
+local http = require("resty.http")
+local ngx  = ngx
+local getenv = os.getenv
+local plugin_name = "azure-functions"
+
+local env_key = {
+    API = "AZURE_FUNCTIONS_APIKEY",
+    CLIENT_ID = "AZURE_FUNCTIONS_CLIENTID"
+}
+
+local schema = {
+    type = "object",
+    properties = {
+        function_uri = {type = "string"},
+        authorization = {
+            type = "object",
+            properties = {
+                apikey = {type = "string"},
+                clientid = {type = "string"}
+            }
+        },
+        timeout = {type = "integer", minimum = 1000, default = 3000},
+        ssl_verify = {type = "boolean", default = true},
+        keepalive = {type = "boolean", default = true},
+        keepalive_timeout = {type = "integer", minimum = 1000, default = 60000},
+        keepalive_pool = {type = "integer", minimum = 1, default = 5}
+    },
+    required = {"function_uri"}
+}
+
+local _M = {
+    version = 0.1,
+    priority = 505,
+    name = plugin_name,
+    schema = schema,
+}
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+function _M.access(conf, ctx)
+    local uri_args = core.request.get_uri_args(ctx)
+    local headers = core.request.headers(ctx) or {}
+    local req_body, _ = core.request.get_body()
+
+
+    -- set authorization headers
+    if conf.authorization then
+        headers["x-functions-key"] = conf.authorization.apikey or ""
+        headers["x-functions-clientid"] = conf.authorization.clientid or ""
+    else
+        headers["x-functions-key"] = getenv(env_key.API)
```

Comment:
We need to expose the AZURE_FUNCTIONS_APIKEY via `env AZURE_FUNCTIONS_APIKEY`, if this plugin is enabled. Nginx will hide all environment variables by default.

Reply:
You mean like this `env AZURE_FUNCTIONS_APIKEY="key" make run`?
Or you are suggesting to update something?

Reply:
Like https://github.com/apache/apisix/blob/2c12b3695302f580a1498333aaafb68da26a3096/apisix/cli/ngx_tpl.lua#L52, but only when this plugin is enabled:
https://github.com/apache/apisix/blob/2c12b3695302f580a1498333aaafb68da26a3096/apisix/cli/ngx_tpl.lua#L209

Reply:
I change my mind. Maybe it could be easier to pass this argument via plugin metadata.

Reply:
Metadata won't be a great fit, IMHO. There might be multiple instances of plugins linked to different routes and we can't force users to use the same key they put into the metadata schema(must be the master key, instead of function-specific keys). This clearly violates principle of Least Access Privileges, isn't it?

Reply:
Removed env vars and added metadata as a fallback option